### PR TITLE
Make hasSelfMapping per tensor.

### PR DESCRIPTION
**Pull request by wujingyue**
For #1768.

### CODE REVIEW

**Comment by wujingyue** [comment]
!build

File: csrc/id_model/id_model.h [hunk 1]
```diff
@@ -121,11 +129,14 @@ class IdModel : public PolymorphicBase {
...
+  // those domains should never be mapped with each other. It may be
+  // possible to lift this assumption, but it's unclear if it could
+  // matter in practice.
+  //
   // TODO: Can we make this more of an alias analysis?
   // Ref: https://github.com/csarofeen/pytorch/pull/1954#discussion_r961940498
-  bool hasSelfMapping() const {
-    return self_mapping_info_.has_value();
-  }
+  std::optional<SelfMapping> hasSelfMapping(const TensorView& tv) const;
```
Comment (naoyam) [inline csrc/id_model/id_model.h:139]: nit: 
```suggestion
  std::optional<SelfMapping> hasSelfMapping(const TensorView* tv) const;
```

While I generally prefer reference types over pointer types whenever possible, the overall codebase pretty much always uses pointer types for historical reasons, and I think it's more important to keep it consistent.

**Review by naoyam — APPROVED** [review]
LGTM

**Comment by wujingyue** [comment]
!build

**Pull request merged**